### ISSUE TITLE
SSplotSexRatio should be called by SS_plots

### ISSUE THREAD

**Issue by taylori** [r4ss/r4ss#286]
The `SSplotSexRatio` function written by @colemonnahan in 2015 wasn't working for SS version 3.30 models. Thanks to Ian Stewart reporting it, this is now working in commit dc0adf1bbacf0dc06c9911b38546769757e28e55 in both the master and development branches (currently they are the same).

In the future, this function should be integrated into the set of default plots created by `SS_plots`, but I don't have time to do that right away.

Until that happens, users can explore the function with commands like the following:
```
# plot of observed and expected sex ratio by year/length bin for fleet 1
SSplotSexRatio(model, kind = "AGE", fleets = 1)

# plot of observed and expected sex ratio by year/age bin for fleet 1
# with additional increase in upper limit of y axis (default is 4)
SSplotSexRatio(model, kind = "LEN", fleets = 1, yupper = 20)

# plots of sex ratios by age and length for all fleets written to PNG files
plotdir <- file.path(model$inputs$dir, "plots")
SSplotSexRatio(model, kind = "AGE", fleets = "all", yupper = 4,
               plot = FALSE, print = TRUE, plotdir = plotdir)
SSplotSexRatio(model, kind = "LEN", fleets = "all", yupper = 2
               plot = FALSE, print = TRUE, plotdir = plotdir)
```

An example of the resulting plots is below:
![sexratio_len_flt1mkt2_page1](https://user-images.githubusercontent.com/4992918/56387600-93fade80-61d9-11e9-88b7-b8a40933ad60.png)

**Comment by iantaylor-NOAA** [comment]
These plots have been slightly revised to show the fraction females rather than females:males and are now integrated into default set created by `SS_plots()` in the `sexratio` branch.

These plots are skipped in combined-sex models as there's nothing to show, but there are likely some model configurations for which these plots might break or be wrong, such as models with multiple seasons within a year. I don't have time for further testing at the moment so I'm leaving users to try them out if they wish by installing the `sexratio` branch (which otherwise matches the current `development` branch).

```
devtools::install_github('r4ss/r4ss', ref='sexratio')
```
If `SS_plots()` crashes during plot group 16 or 17, go back to the main trunk of r4ss or the `development` branch via 
```
devtools::install_github('r4ss/r4ss', ref='development')
```

The plots will appear in the "LenComp" and "AgeComp" tabs of the HTML output created by `SS_plots()` as shown in the example below showing the results for a 2015 Petrale Sole model.

![image](https://user-images.githubusercontent.com/4992918/57261400-7ffa0e00-701c-11e9-8db5-c40f2457d96c.png)

**Comment by kellijohnson-NOAA** [comment]
@iantaylor-NOAA thank you for these plots. Jim Ianelli was asking for something like this during the sablefish STAR panel but I didn't know that there was already code for them so his request was not fulfilled. I switched branches and generated them for sablefish from a V3.30.13-safe run. I have some questions.
 * Should we place each plot next to other plots for that fleet x composition type, move them to a separate tab, or keep them as is, where you have to scroll to the bottom of the page? I actually like where they are, but this is a consistency / fundamental question if composition plots in general should be aggregated by type or by fleet. I am hesitant to mix methods and to change the current method.
 * Can we add a reference for "Jeffreys interval"? I googled it, which led me down a rabbit hole.
 * The sexratio plots were not created for the fleet with conditional-age-at-length.
 * A "main" label would be nice so that when you click on the plot you still know what fleet you are looking at.
 * Should the fit line be red? I know that all other fit lines are blue in r4ss plots if only one sex is shown, but it might be confusing to have a blue line when you are displaying the fit to fraction female rather than fraction male given that males and females are coloured blue and red in all other plots.

**Comment by iantaylor-NOAA** [comment]
Hi @kellijohnson-NOAA. These plots were implemented in the `sexratio` branch during the skate STAR panel and I didn't have time to test them against a wider range of models so didn't merge them in. Sorry I should have spread the word better on their availability in a branch. And thanks for all the good suggestions.

Regarding your suggestions,

- *plot ordering*: The sex ratio plots are really just a different way to display the composition fit, so that's why they're on the same tab in the HTML output, but I didn't put any thought into the order they occur there. Currently all the other plots on that tab are created by the `SSplotComps()` function, whereas the sex ratio plots are created by `SSplotSexRatio()`. To group them with the other plots for each fleet it would probably be easiest to have `SSplotComps()` call `SSplotSexRatio()` which wouldn't be too hard. I also would like to move the plots showing comps aggregated across all years for all fleets to the top of the tab rather than have them below the fleet-specific plots.

- *add reference*: I also thought I had added a reference to Brown, L.D., Cai, T.T. and DasGupta, A., 2001. Interval estimation for a binomial proportion. Statistical science, pp.101-117 for the Jeffreys interval but now see that it was in the Big Skate report, not the r4ss plots. The Francis weighting plots are each followed by a reference to Francis (2011) and the comp fits using Dirichlet-Multinomial weighting are each followed by a reference to Thorson et al. (2017). I can follow the same approach with the sex ratio plots. However, would it be good to instead add these references only after the first plot of each type? The reference should also be added to the help page for SSplotSexRatio but since users probably won't call that function directly, it would be less useful there.

- *Conditional-age-at-length* There are two things going on here. First, the `SSplotSexRatio()` function only works for marginal composition data, not conditional data. Second, conditional age-at-length data is typically entered separately for females and males with sex = 1 and 2 respectively, rather than sex = 3 which causes Stock Syntheses to interpret them as independent vectors that sum to 1.0 in each case which are fit to a model expectation that also sums to 1.0 in each case regardless of the estimated sex ratio within that length bin. In theory `SSplotSexRatio()` works for marginal data which have been entered separately as sex = 1 and 2 but I haven't tested it against a model with data input that way and I think it should probably be disabled since the choice to enter the sexes separately is typically made to explicitly remove any assumptions about sex ratios. Having said all that, it might be useful to expand the plot in the future to work with conditional age-at-length data if they use sex = 3.

- *add "main" label* This is related to long-standing partially-implemented issue #6, which is to remove all main labels when writing to PNG files with HTML captions. Ian Stewart used to (maybe still does) manually crop the main label in Word for every figure in an assessment report. When you click on the plot, the URL shown in your browser should include the filename which includes the fleet number. If you run `SS_plots(..., mainTitle = TRUE)` you should get the main labels on the sex ratio plots (and many of the other comp plots which currently don't have them).

- *color of fit line* The comp fits in this tab use red for expected values for females, blue for males, and green for unsexed fish, so perhaps green should be used here. Although the y-axis says "Fraction female", the "Fraction male" plot would just be the mirror image of the same plot, so I think it makes sense to think of this as representative of both sexes. Or how about purple?

In summary, here's a (perhaps incomplete) to-do list:

- [ ] decide on and implement a plot ordering scheme (which could be status-quo)
- [ ] add a reference (decide how many times)
- [ ] consider disabling sex ratio plots for models with marginal comps entered as sex = 1 or 2
- [ ] consider adding a separate issue for the future addition of sex ratio plots for conditional data entered as sex = 3 (a low priority in my opinion)
- [ ] change the color of the fit lines to whatever we decide
- [ ] test the `sexratio` branch on a range of models
- [ ] merge the `sexratio` branch into `development`

**Comment by chantelwetzel-noaa** [comment]
I don't have much to add here because I think the main issues have been well covered by both of you.  However, I would suggest that we could deal with the CAAL data issue is to ensure that the code works with "ghost" fleets which would provide a way for users to visualize these data when input as marginals for a ghost fleet.

**Comment by iantaylor-NOAA** [comment]
Good point @chantelwetzel-noaa. For Big Skate, I changed the marginal age comps from a ghost fleet to a real fleet with input sample size 0.001 in order to get the sex ratio plots, but then turned that off for the final model.
- [ ] plot sex ratios for "ghost" marginal comps

**Comment by kellijohnson-NOAA** [comment]
I have a suggestion regarding the references. I think that we should hyperlink references in the caption that sends users to the journal article and only list the full reference either at the bottom of the relevant html page or in a new page that is strictly for references. 
E.g., 
-Image-
Mean age for Fleet 1 with 95% confidence intervals based on current samples sizes.<br><a href = 'https://doi.org/10.1139/f2011-025' target = '_blank'>Francis data weighting method TA1.8</a>: thinner intervals (with capped ends) show result of further adjusting sample sizes based on suggested multiplier (with 95% interval) for age data from Fleet 1:<br>1.8607 (1.1705-3.6731)<br><i><small>file: <a href='comp_agefit_flt3mkt2_page1.png'>comp_agefit_flt1mkt2_page1.png</a></small></i>
-More images-
...

References
Francis, R.I.C.C. (2011). Data weighting in statistical fisheries stock assessment models. Can. J. Fish. Aquat. Sci. 68: 1124-1138. https://doi.org/10.1139/f2011-025
Thorson, J.T., Johnson, K.F., Methot, R.D. and Taylor, I.G. 2017. Model-based estimates of effective sample size in stock assessment models using the Dirichlet-multinomial distribution. Fisheries Research192: 84-93. https://doi.org/10.1016/j.fishres.2016.06.005

**Comment by iantaylor-NOAA** [comment]
In commit df9980301240aab2ca68ec1ade98be396020b635 I made two easy changes: line color from blue to purple and linkified "[Jeffreys interval](https://www.jstor.org/stable/2676784)" in the captions (though I like the idea of the reference section at the bottom of the file as an option proposed by @kellijohnson-NOAA above).

The `sexratio` branch is now merged into `development` so these plots will be more widely seen. The branch will remain available for future progress as discussed under this issue.
